larray/core/constrained.py

```diff
@@ -0,0 +1,533 @@
+from abc import ABCMeta
+from copy import deepcopy
+import warnings
+
+import numpy as np
+
+from typing import TYPE_CHECKING, Type, Any, Dict, Set, List, no_type_check
+
+from larray.core.metadata import Metadata
+from larray.core.axis import AxisCollection
+from larray.core.array import Array, empty
+from larray.core.session import Session
+
+
+class NotLoaded:
+    pass
+
+
+try:
+    import pydantic
+except ImportError:
+    pydantic = None
+
+if pydantic:
+    from pydantic.fields import ModelField
+    from pydantic.class_validators import Validator
+    from pydantic.main import BaseConfig
+
+    # the implementation of the class below is inspired by the 'ConstrainedBytes' class
+    # from the types.py module of the 'pydantic' library
+    class ConstrainedArrayImpl(Array):
+        expected_axes: AxisCollection
+        dtype: np.dtype = np.dtype(float)
+
+        # see https://pydantic-docs.helpmanual.io/usage/types/#classes-with-__get_validators__
+        @classmethod
+        def __get_validators__(cls):
+            # one or more validators may be yielded which will be called in the
+            # order to validate the input, each validator will receive as an input
+            # the value returned from the previous validator
+            yield cls.validate
+
+        @classmethod
+        def validate(cls, value, field: ModelField):
+            # XXX: not sure we should accept a scalar as value but this allows to not repeat axes
```

Comment:
Is this comment still valid? scalar as values seem fine to me.

Reply:
I guess not. Since I allowed broadcasting, I don't see any reason to not accept scalar.